tests/TestCase/Validation/ValidationSetTest.php

```diff
@@ -215,12 +215,12 @@ public function testAllowEmpty(): void
     {
         $set = new ValidationSet();
 
-        $this->assertFalse($set->isEmptyAllowed());
-
-        $set->allowEmpty(true);
         $this->assertTrue($set->isEmptyAllowed());
 
         $set->allowEmpty(false);
         $this->assertFalse($set->isEmptyAllowed());
+
+        $set->allowEmpty(true);
+        $this->assertTrue($set->isEmptyAllowed());
```

Comment:
@andrii-pukhalevych your proposed change not only addresses the unwanted/unexpected behavior of the ``FormHelper``, generating form fields with `required=required` set where you don't want them, but changes how the whole CakePHP framework handles field emptiness by default.

While previously, fields were not allowed to be empty by default, with your rchange they could now.

I don't think this is the proper way to address the issue at hand because this will likely have significant impact on all CakePHP applications. Even if we introduce it in a major release where backwards incompatible changes are allowed.

A better way to tackle this issue was mentioned by @markstory in [his comment](https://github.com/cakephp/cakephp/issues/15699#issuecomment-1033857410):

> We could update FormHelper to detect whether or not a field is required to be non-empty so that the required attribute is only applied when the field is both required and has a validation rule or not-empty requirement.

So let's try to fix a frontend inconsistency in the frontend code of CakePHP.

Two possible candidates for a frontend-fix would be these:

https://api.cakephp.org/5.0/class-Cake.View.Form.EntityContext.html#isRequired()
https://api.cakephp.org/5.0/class-Cake.View.Form.FormContext.html#isRequired()

Reply:
So the issue occurs because of how `Validator::isEmpty()` works. Since no `allowEmpty*/notEmpty*()` call was used for the field, `validate()` calls `isEmpty()` with flag `Validator::EMPTY_NULL` (and value `''`) causing it to return `false`. Hence empty string gets allowed and validation passes.

_Originally posted by @ADmad in https://github.com/cakephp/cakephp/issues/15699#issuecomment-932162579_

Reply:
Should `Validator::isEmpty()` return `true` with flag `Validator::EMPTY_NULL` and value ''?
Otherwise, "fields were not allowed to be empty by default" rule does not work and its backend bug, not in frontend.